### PR TITLE
Ports - Margaret

### PR DESCRIPTION
**Pull request by msfinnan**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | How to create reservations and then book hotel blocks.
What was a design decision you made that changed over time over the project? | At first, I had a complicated make_reservation class that reserved regular rooms and rooms from blocks. Eventually I split this into two methods and it made each method have a single job and made the logic much easier. 
What was a concept you gained clarity on, or a learning that you'd like to share? | I gained much more clarity on the role of attr_readers and instance variables. I got comfortable writing tests and thinking of things to test for. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | For the Reservation duration method I tested that it can calculate the length of a stay between 2 dates in the future. This is nominal because this is what the method is intended to do. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | For the same method I created a test case that tested if it would raise an error if the trip was negative days (the end date was before the start date). This is an edge case because it is an unlikely scenario, but accounts for possible user input errors. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | It helped me have a very clear understanding of the function and requirements of the program. The times I thoroughly pseudocoded I was able to make the code work much faster than when I code without it. I also was able to make the code work in ways I thought I didn't know how to make it work. I'm definitely still getting used to pseudocoding. I really like writing tests before I write the code so that I know right away if it is working the way I expect it to or not.

### CODE REVIEW

File: lib/reservation_manager.rb [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      @blocked_reservation_array = []
+      @rooms = ("1".."20").to_a
+    end
+
+    def make_reservation(start_date: Date.today.to_s, end_date: (Date.today + 1).to_s, room: "0", cost: 200, block_name: "no name")
+      new_reservation = Reservation.new(start_date: start_date, end_date: end_date, room: room, cost: cost, block_name: block_name)
+      check_res = view_available_rooms(start_date: new_reservation.start_date, end_date: new_reservation.end_date)
+      if !check_res.include?(new_reservation.room)
+        raise ArgumentError, "That room is not available, choose another room"
+      elsif new_reservation.block_name != "no name"
```
Comment (CheezItMan) [inline lib/reservation_manager.rb]: I would suggest that `block_name` default to `nil` instead of `"no name"` just to keep someone from naming their block `"no name"`.

File: lib/reservation_manager.rb [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        raise ArgumentError, "Use the make_block_reservation method to reserve a room from a block"
+      else
+        @reservation_array << new_reservation
+      end
+      return new_reservation
+    end
+
+    def make_block_reservation(start_date: Date.today.to_s, end_date: (Date.today + 1).to_s, room: "0", cost: 200, block_name: "no name")
```
Comment (CheezItMan) [inline lib/reservation_manager.rb]: Why provide default dates for the reservation if the reservation is required to have the same dates as the block, which are unlikely to be the same as the defaults.

File: lib/reservation_manager.rb [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+      @blocked_reservation_array.each do |blocked_res|
+        if blocked_res.reservation_dates != new_reservation.reservation_dates && blocked_res.block_name == new_reservation.block_name
+          raise ArgumentError, "You must reserve a blocked hotel room for its entire duration"
+        end
+      end
+
+      @reservation_array << new_reservation
+
+      res_to_remove = []
```
Comment (CheezItMan) [inline lib/reservation_manager.rb]: why make this an array if you are only going to get the 1st element.

File: lib/reservation_manager.rb [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            if duration[0..-2].include?(date) && blocked_reservation.room.include?(room)
+              raise ArgumentError, "There is already a block on that date or room during your block duration."
+            end
+          end
+        end
+      end
+
+      rooms_array.each do |block_room|
+        block_reservation = Reservation.new(start_date: start_date, end_date: end_date, room: block_room, cost: cost, block_name: block_name)
+        @blocked_rooms_array << block_reservation.room
```
Comment (CheezItMan) [inline lib/reservation_manager.rb]: Instead of making parallel arrays like this, maybe just use the `@blocked_reservation_array`.

File: lib/reservation_manager.rb [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+      date_range[0..-2].each do |date|
+        @reservation_array.each do |reservation|
+          if reservation.reservation_dates[0..-2].include?(date)
+            @booked_rooms << reservation.room
+          end
+        end
+      end
+
+      @available_rooms = rooms - @booked_rooms
```
Comment (CheezItMan) [inline lib/reservation_manager.rb:124]: Good use of array methods

File: spec/reservation_manager_spec.rb [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    it "Allows the user to reserve a room for a given date range" do
+      reservation_manager.make_reservation(start_date: "1st Jan 2019", end_date: "3rd Jan 2019", room: "1")
+      expect(reservation_manager.reservation_array[0].room).must_equal "1"
+    end
+    it "Allows user to reserve an available room for a given date range" do
+      reservation_manager.make_reservation(start_date: "1st Jan 2019", end_date: "3rd Jan 2019", room: "1")
+      reservation_manager.make_reservation(start_date: "1st Jan 2019", end_date: "3rd Jan 2019", room: "2")
+      expect(reservation_manager.reservation_array[1].room).must_equal "2"
+    end
+    it "Allows user to reserve an available room for a given date range (starts on the checkout date)" do
```
Comment (CheezItMan) [inline spec/reservation_manager_spec.rb:32]: Good checking of edge-cases

File: spec/reservation_manager_spec.rb [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    end
+    it "Raises an ArgumentError if the user tries to reserve a room that is unavailable for a given date range (overlaps one day)" do
+      reservation_manager.make_reservation(start_date: "1st Jan 2019", end_date: "3rd Jan 2019", room: "1")
+      expect { reservation_manager.make_reservation(start_date: "2nd Jan 2019", end_date: "4th Jan 2019", room: "1") }.must_raise ArgumentError
+    end
+    it "Raises an ArgumentError if the user tries to reserve a room that is unavailable for a given date range (completely contained)" do
+      reservation_manager.make_reservation(start_date: "1st Jan 2019", end_date: "5th Jan 2019", room: "1")
+      expect { reservation_manager.make_reservation(start_date: "2nd Jan 2019", end_date: "4th Jan 2019", room: "1") }.must_raise ArgumentError
+    end
+    it "Raises an ArgumentError if the user tries to reserve a room that is unavailable for a given date range (completely containing)" do
```
Comment (CheezItMan) [inline spec/reservation_manager_spec.rb]: This test name doesn't seem right

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | Good number of commits and good commit messages
Answer comprehension questions | Check, 
**Design** |
Each class is responsible for a single piece of the program | Check, well done
Classes are loosely coupled | Check
**Wave 1** |
List rooms | Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check
Invalid date range produces an error | Check
**Wave 2** |
View available rooms for a given date range | Check
Reserving a room that is not available produces an error | Check
**Wave 3** |
Create a block of rooms | Check
Check if a block has rooms | Check
Reserve a room from a block | Check
Test coverage | 100% nice!
**Fundamentals** |
Names variables, classes and modules appropriately | Check
Understanding of variable scope - local vs instance | Check
Can create complex logical structures utilizing variables | Check
Appropriately uses methods to break down tasks into smaller simpler tasks | Check
Understands the differences between class and instance methods | Check
Appropriately uses iterators and `Enumerable` methods | Check
Appropriately writes and utilizes classes | Check
Appropriately utilizes modules as a namespace | Check
**Wrap Up** |
There is a refactors.txt file | Check
The file provides a roadmap to future changes | Check
**Additional Feedback** | Well done, you hit all the learning goals and wrote some excellent tests.  Great work!  Check out my comments let me know if you need anything explained more throughly.
